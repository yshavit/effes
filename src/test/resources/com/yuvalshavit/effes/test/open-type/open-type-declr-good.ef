type Pet ?
  isHappy (fed: True | False) -> True | False

type Cat:
  is Pet
  isHappy (fed: True | False) -> True | False: return fed

type Dog:
  is Pet
  isHappy (fed: True | False | Void) -> True: return True -- note contravariant args, covariant return type

main:
  print Cat.isHappy True
  print Cat.isHappy False
  print Dog.isHappy Void
