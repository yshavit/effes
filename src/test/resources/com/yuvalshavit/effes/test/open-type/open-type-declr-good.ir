def Cat::isHappy (fed: False | True) -> False | True:
  ReturnStatement
    VarExpression
      arg: fed
      pos: 1
def Dog::isHappy (fed: False | True | Void) -> True:
  ReturnStatement
    CtorInvoke
      type: True
