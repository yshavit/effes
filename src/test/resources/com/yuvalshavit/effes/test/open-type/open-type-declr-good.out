Your argument was: True
Your argument was: False
Your argument was: True
>> null
